solidity/core/contracts/Outbox.sol

```diff
@@ -116,14 +110,10 @@ contract Outbox is IOutbox, Version0, MerkleTreeManager, Common {
         bytes memory _messageBody
     ) external override notFailed returns (uint256) {
         require(_messageBody.length <= MAX_MESSAGE_BODY_BYTES, "msg too long");
-        // get the next nonce for the destination domain, then increment it
-        uint32 _nonce = nonces[_destinationDomain];
-        nonces[_destinationDomain] = _nonce + 1;
         // format the message into packed bytes
         bytes memory _message = Message.formatMessage(
```

Comment:
Should we be just replacing `_nonce` with `_leafIndex` here? Iirc we still need something nonce-like to ensure there are no message hash conflicts (important because in Inbox.sol we rely on `messages[_messageHash]` for determining if a message has been processed or not), but we just don't need a domain-specific nonce because the leaf index is sufficient to avoid message hash conflicts

Reply:
Like discussed on Discord, great point. I ultimately decided to go with explicitly passing the leaf_index to the message hash calculation instead of embedding the leaf_index inside the message, because the duplication felt very wrong, and most places that deal with messages retrieve them from the dispatch even where they will already have access to the leafindex